src/validation.cpp

```diff
@@ -4738,6 +4748,46 @@ bool GetBlockPublicKey(const CBlock& block, std::vector<unsigned char>& vchPubKe
     return false;
 }
 
+bool GetBlockDelegation(const CBlock& block, const uint160& staker, uint160& address, uint8_t& fee)
+{
+    // Check block parameters
+    if (block.IsProofOfWork())
+        return false;
+
+    if (block.vchBlockSigDlgt.empty())
+        return false;
+
+    if (!block.HasDelegation())
+        return false;
+
+    // Get the delegate
+    std::string strMessage = staker.ToString();
+    CKeyID keyid;
+    if(!SignStr::GetKeyIdMessage(strMessage, block.GetBlockDelegate(), keyid))
+        return false;
+    address = uint160(keyid);
+
+    // Get the staker fee
+    CCoinsViewCache& cache = ::ChainstateActive().CoinsTip();
+    COutPoint prevout = block.vtx[1]->vin[0].prevout;
+    if(!cache.HaveCoinInCache(prevout))
+        return false;
+
+    CAmount nValueCoin = cache.AccessCoin(prevout).out.nValue;
+    if(nValueCoin <= 0)
+        return false;
+
+    CAmount nValueStaker = block.vtx[1]->vout[0].nValue;
+    CAmount nValueDelegate = block.vtx[1]->vout[1].nValue;
+    CAmount nReward = nValueStaker + nValueDelegate - nValueCoin;
+    if(nReward < 0)
+        return false;
+
+    fee = std::lround(nValueStaker * 100.0 / nReward);
```

Comment:
`nValueStaker = nReward * fee / 100`, so `fee = std::ceill(nValueStaker * 100.0 / nReward)`

Reply:
And it may produce a `0.0/0.0` when block subsidy reduces to zero

Reply:
Fixed